buildSrc/src/main/kotlin/BuildInfo.kt

```diff
@@ -80,6 +127,219 @@ open class BuildInfo(project: Project) {
+            // multiply out by jdk vendor
+            testJdkVendors.map { vendor -> (targetVersion to vendor) }
+          }
+          .filter { (jdkTarget, vendor) ->
+            // only include experimental tasks in the return suite if the flag is set. if the task
+            // is withheld from the returned list, it will not be executed by default with `gradle
+            // check`.
+            testExperimentalJdks ||
+              (!namer(jdkTarget, vendor.takeIf { isMultiVendor }).contains("Experimental"))
+          }
+          .map { (jdkTarget, vendor) ->
+            if (jdkToolchainVersion == jdkTarget)
+              tasks.register(namer(jdkTarget, vendor)) {
+                // alias to `test`
+                dependsOn(templateTask)
+                group = Category.VERIFICATION
+                description =
+                  "Alias for regular '${baseNameProvider()}' task, on JDK ${jdkTarget.asInt()}"
+              }
+            else
+              the<TestingExtension>().suites.register(
+                namer(jdkTarget, vendor.takeIf { isMultiVendor }),
+                JvmTestSuite::class,
+              ) {
+                targets.all {
+                  testTask.configure {
+                    group = Category.VERIFICATION
+                    description = "Run tests against JDK ${jdkTarget.asInt()}"
+                    applyConfig(jdkTarget to toolchains.launcherFor { languageVersion = jdkTarget })
+
+                    // fix: on jdk17, we must force the polyglot module on to the modulepath
+                    if (jdkTarget.asInt() == 17)
+                      jvmArgumentProviders.add(
+                        CommandLineArgumentProvider {
+                          buildList { listOf("--add-modules=org.graalvm.polyglot") }
+                        }
+                      )
+                  }
+                }
+              }
+          }
+          .toList()
+      }
+    }
+
+  val javaCompiler: Provider<JavaCompiler> by lazy {
+    project.serviceOf<JavaToolchainService>().let { toolchainService ->
+      toolchainService.compilerFor { pklJdkToolchain() }
+    }
+  }
+
+  val javaTestLauncher: Provider<JavaLauncher> by lazy {
+    project.serviceOf<JavaToolchainService>().let { toolchainService ->
+      toolchainService.launcherFor { pklJdkToolchain() }
+    }
+  }
+
+  val multiJdkTesting: Boolean by lazy {
+    // By default, Pkl is tested against a full range of JDK versions, past and present, within the
+    // supported bounds of `PKL_TEST_JDK_TARGET` and `PKL_TEST_JDK_MAXIMUM`. To opt-out of this
+    // behavior, set `-DpklMultiJdkTesting=false` on the Gradle command line.
+    //
+    // In CI, this defaults to `true` to catch potential cross-JDK compat regressions or other bugs.
+    // In local dev, this defaults to `false` to speed up the build and reduce contributor load.
+    System.getProperty("pklMultiJdkTesting")?.toBoolean() ?: isCiBuild
+  }
```

Comment:
Instead of disabling `pklMultiJdkTesting` by default, I've kept it enabled in CI, and disabled by default in local dev.